### PR TITLE
Initial commit adding dependabot.

### LINKED ISSUE #67: Unpin Python bindings for Glean SDK 📦
See https://github.com/mozilla/burnham/blob/main/application/setup.py#L36

### LINKED ISSUE #71: Enable auto pull-requests for new Glean SDK releases 📦
We unpinned the Glean SDK version in #67. The next step is to automatically build and push an updated burnham Docker image to GKE for new releases of the Python bindings on PyPI. We have configured CircleCI to build and push Docker images for pushes to the main branch. Can we add a requirements.txt file and set up Dependabot to automatically open pull-requests against that file for new releases of the Glean SDK? 🤖 

https://github.blog/2020-06-01-keep-all-your-packages-up-to-date-with-dependabot/

### PR DESCRIPTION
**Pull request by jrbenny35**
Closes #71

### CODE REVIEW

File: dependabot.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+version: 2
+updates:
+  - package-ecosystem: "pip"
+    directory: "/"
+    schedule:
+      interval: "weekly"
```
Comment (jklukas) [inline dependabot.yml:6]: We may want to make this daily to shorten the time between a new glean sdk version being published and it being exercised by burnham.

**Review by jklukas — COMMENTED** [review]
I think the scope of this PR should include using the requirements.txt in `application/Dockerfile` for determining which specific dependencies are included in the published docker container. Or would you prefer to break that out to a separate PR?

This looks good except that I was expecting the `requirements.txt` to appear under the `application` folder. This repo contains two separate logical applications (application/ and bigquery/), and I think we want to keep their dependencies separate.

**Comment by jrbenny35** [comment]
> I think the scope of this PR should include using the requirements.txt in `application/Dockerfile` for determining which specific dependencies are included in the published docker container. Or would you prefer to break that out to a separate PR?
> 

These requirements were taken from the 4 in ```applications/setup.py``` the ```install_requires```. I can move the file over though, I agree it should not exist at ```/```

File: dependabot.yml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+version: 2
+updates:
+  - package-ecosystem: "pip"
+    directory: "/"
```
Comment (hackebrot) [inline dependabot.yml:4]: Do we need to update this to point to `application/` now that you moved the file?

**Review by hackebrot — COMMENTED** [review]
I agree with @jklukas, we can include updating `application/Docker` to install from `application/requirements.txt` in this pull-request. 🤖

File: application/setup.py [hunk 3]
```diff
@@ -33,7 +33,7 @@ def read(*args: str) -> str:
     package_data={"burnham": ["config/*.yaml"]},
     zip_safe=False,
     python_requires=">=3.7",
-    install_requires=["click>=7.0", "glean-sdk>=32.3.1", "wrapt", "typing_extensions"],
+    install_requires=[],
```
Comment (hackebrot) [inline application/setup.py:36]: We still need to define direct dependencies for burnham here.

File: dependabot.yml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+version: 2
+updates:
+  - package-ecosystem: "pip"
+    directory: "application/"
+    schedule:
+      interval: "weekly"
```
Comment (hackebrot) [inline dependabot.yml:6]: Please update the `interval` as per @jklukas's suggestion.

**Review by hackebrot — CHANGES_REQUESTED** [review]
Please modify `application/Docker` to create wheels from `application/requirements.txt` and make sure the correct packages are installed when creating the container.

File: application/setup.py (hunk 3 above)
Comment (jklukas) [reply application/setup.py:36]: It does get kind of confusing what to put where when you have both `install_requires` and `requirements.txt`. What I've seen is that `install_requires` defines minimal dependencies without pinning versions, so that if another project were importing this as a library, it would be flexible. And then requirements.txt has everything pinned for maximum stability of CI and published docker images.

So I think this line can stay as it currently exists in `main`.

Comment (hackebrot) [reply application/setup.py:36]: > So I think this line can stay as it currently exists in main.

Yes, I agree. 👍

Comment (jrbenny35) [reply application/setup.py:36]: I'm not sure I follow. I thought if you have a requirements file you don't need to specify dependencies here.

Comment (jklukas) [reply application/setup.py:36]: > I thought if you have a requirements file you don't need to specify dependencies here.

That is technically true in this case because `burnham` isn't published as a library. It's still nice to include `install_requires` as a form of documentation, though, as it captures the logical dependencies without the noise of the transitive dependencies and pinned versions.

That said, there is the downside that we could well end up adding new dependencies in requirements and forget to update the list here, so it could become stale.

Comment (hackebrot) [reply application/setup.py:36]: We list direct dependencies for burnham in `install_requires` because we know that burnham cannot work without these packages. We import from these specific packages and rely on the APIs to be available.

```python
install_requires=["click>=7.0", "glean-sdk>=32.3.1", "wrapt", "typing_extensions"]
```

We don't necessarily specify a version range for the direct dependencies, as long as we know that a certain feature became available in a specific version or that an API we rely on is deprecated and will be removed in a future release of that dependency. We also don't specify the dependencies of direct dependencies, because those are internal details of our dependencies.

The purpose of pinning versions for direct dependencies and transitive dependencies in a `requirements.txt` file is reproducibility (aside from making it easier to integrate Dependabot). When we install from this file, we know exactly which packages are installed and given that we pin dependencies, we also know which version is installed when we run the burnham application.

We expect the packages listed as dependencies in `install_requires` to be available after installing from `requirements.txt`. If we don't specify `install_requires`, we lose the error reporting at the installation stage and will only find out when running burnham that we are missing a dependency.

In summary, `install_requires` is important for installing the burnham package and its direct dependencies and `requirements.txt` is important for making sure we know exactly which packages and versions are installed when building the burnham Docker image and running that in our test runs.

Hope this helps!

File: application/Dockerfile [hunk 5]
```diff
@@ -9,7 +9,7 @@ RUN python -m pip install --upgrade --no-cache-dir \
     pip>=20.1.1 \
     setuptools>=49.1.0 \
     wheel>=0.34.2
-RUN python -m pip wheel -w /wheels .
+RUN python -m pip wheel -w /wheels -r requirements.txt
```
Comment (jklukas) [inline application/Dockerfile:12]:
```suggestion
RUN python -m pip wheel -w /wheels -r requirements.txt
RUN python -m pip wheel -w /wheels .
```

I believe we still need to instruct `pip` to install the local package. It's possible we could do this all in one invocation, or by adding `.` to the `requirements.txt` file, but I think two separate invocations is the clearest way to express the intent.

I think this is the only remaining detail. Otherwise, this PR is looking great!

Comment (jrbenny35) [reply application/Dockerfile:12]: Ah yes, thanks for pointing this out

**Review by jklukas — APPROVED** [review]
r+ with comment

This just needs to have the `RUN python -m pip wheel -w /wheels .` line added. You can either do that locally and push your changes, or you can use the "Commit suggestion" dropdown on https://github.com/mozilla/burnham/pull/78/files#r499831296.

**Comment by jrbenny35** [comment]
> r+ with comment
> 
> This just needs to have the `RUN python -m pip wheel -w /wheels .` line added. You can either do that locally and push your changes, or you can use the "Commit suggestion" dropdown on https://github.com/mozilla/burnham/pull/78/files#r499831296.

I added that. Can you double-check? I can see it

**Comment by hackebrot** [comment]
The `RUN python -m pip wheel -w /wheels .` line doesn't show as an addition in the diff because it's already on the `main` branch. I can see it in the file on Ben's branch.

**Comment by hackebrot** [comment]
I was chatting with @jezdez about pip wheel today and he suggested to combine the two `RUN python -m pip wheel -w /wheels` steps into one by using only `RUN python -m pip wheel -w /wheels -r requirements.txt` and adding `.` to `requirements.txt`. This has the benefit of allowing pip to resolve to one dependency tree and generating wheels in one step. I think that makes a lot of sense and suggest we follow this approach! 📦

File: application/Dockerfile [hunk 6]
```diff
@@ -9,6 +9,7 @@ RUN python -m pip install --upgrade --no-cache-dir \
     pip>=20.1.1 \
     setuptools>=49.1.0 \
     wheel>=0.34.2
+RUN python -m pip wheel -w /wheels -r requirements.txt
```
Comment (jklukas) [inline application/Dockerfile:12]: Per @jezdez suggestion in https://github.com/mozilla/burnham/pull/78#issuecomment-704132772
```suggestion
RUN python -m pip wheel -w /wheels -r requirements.txt .
```

File: application/Dockerfile [hunk 7]
```diff
@@ -9,6 +9,7 @@ RUN python -m pip install --upgrade --no-cache-dir \
     pip>=20.1.1 \
     setuptools>=49.1.0 \
     wheel>=0.34.2
+RUN python -m pip wheel -w /wheels -r requirements.txt .
 RUN python -m pip wheel -w /wheels .
```
Comment (jklukas) [inline application/Dockerfile:13]:
```suggestion
RUN python -m pip wheel -w /wheels -r requirements.txt .
```

File: application/Dockerfile [hunk 8]
```diff
@@ -9,6 +9,7 @@ RUN python -m pip install --upgrade --no-cache-dir \
     pip>=20.1.1 \
     setuptools>=49.1.0 \
     wheel>=0.34.2
+RUN python -m pip wheel -w /wheels -r requirements.txt
 RUN python -m pip wheel -w /wheels .
```
Comment (jklukas) [inline application/Dockerfile:13]:
```suggestion
RUN python -m pip wheel -w /wheels -r requirements.txt .
```

**Comment by jklukas** [comment]
Sorry. I inadvertently committed a bad suggestion to this branch. I then backed it out.

Once we commit the suggestion to go down to a single `pip` invocation, this is good to merge.

File: application/Dockerfile (hunk 8 above)
Comment (jrbenny35) [reply application/Dockerfile:13]: This is what I had originally but it continues the ```-r``` and tries installing the dir as a requirement.

Comment (hackebrot) [reply application/Dockerfile:13]: I’m getting an error for `RUN python -m pip wheel -w /wheels -r requirements.txt .` when I try this locally:

```text
ERROR: Can’t verify hashes for these file:// requirements because they point to directories:
    file:///usr/src/burnham
```

Looks like we cannot combine a requirements file that contains hashes and a local directory.

I suggest we use the following two steps:

```text
RUN python -m pip wheel -w /wheels -r requirements.txt
RUN python -m pip wheel -w /wheels —no-deps .
```

This way we install the packages in the requirements file with hashes and build wheels for them and then build a wheel only for burnham in the next step.

File: dependabot.yml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+version: 2
+updates:
+  - package-ecosystem: "pip"
+    directory: "application/"
+    schedule:
+      interval: "daily"
```
Comment (hackebrot) [inline dependabot.yml:6]: Please add the newline at the end of the file.

File: dependabot.yml [hunk 10]
```diff
@@ -0,0 +1,6 @@
+version: 2
```
Comment (hackebrot) [inline dependabot.yml:1]: Do we need to enable Dependabot in the repo settings? Or is adding this file enough to enable it?

File: application/requirements.txt [hunk 11]
```diff
@@ -0,0 +1,42 @@
+appdirs==1.4.4 \
```
Comment (hackebrot) [inline application/requirements.txt:1]: Which tool did you use to generate this file?

**Review by hackebrot — CHANGES_REQUESTED** [review]
Added a few comments. 📝

File: application/requirements.txt (hunk 11 above)
Comment (jezdez) [reply application/requirements.txt:1]: I recommend using pip-tools' pip-compile for generating this file, which has been used in Mozilla for a while and [recommended elsewhere](https://hynek.me/articles/python-app-deps-2018/), too.

File: application/Dockerfile (hunk 8 above)
Comment (jezdez) [reply application/Dockerfile:13]: Oh wait, that wasn't my recommendation! It was to put `.` **inside** the requirements.txt and not have separate calls to `pip wheel`. pip does the right thing and then first resolves the dependency chain of the `burnham` package in addition to those requirements listed in the requirements file.

The call would just be 

```suggestion
RUN python -m pip wheel --wheel-dir /wheels -r requirements.txt
```

Comment (jezdez) [reply application/Dockerfile:13]: Hm, unless that doesn't work for requirements files with hashes as @hackebrot said in https://github.com/mozilla/burnham/pull/78#discussion_r500800987. So calling it twice is probably the easiest to do this, sadly.

File: application/requirements.txt (hunk 11 above)
Comment (jrbenny35) [reply application/requirements.txt:1]: I didn't use any tool. Every tool I have used has been annoying and to me it's easier just to copy paste but I will try @jezdez suggestion. Should I re-do them?

File: dependabot.yml (hunk 10 above)
Comment (jrbenny35) [reply dependabot.yml:1]: Github docs say ```You enable GitHub Dependabot version updates by checking a configuration file in to your repository. ```, https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/about-github-dependabot-version-updates#about-github-dependabot-version-updates

**Review by hackebrot — APPROVED** [review]
LGTM! 🚀
